plugins/lttng-utils/debug-info.c

```diff
@@ -726,22 +884,1100 @@ void debug_info_handle_event(FILE *err, const bt_event *event,
+	if (!debug_info_comp) {
+		BT_LOGE_STR("Failed to allocate one debug_info component.");
+		goto error;
+	}
+
+	bt_self_component_set_data(
+			bt_self_component_filter_as_self_component(self_comp),
+			debug_info_comp);
+
+	status = bt_self_component_filter_add_input_port(self_comp, "in",
+			NULL, NULL);
+	if (status != BT_SELF_COMPONENT_STATUS_OK) {
+		goto error;
+	}
+
+	status = bt_self_component_filter_add_output_port(self_comp, "out",
+			NULL, NULL);
+	if (status != BT_SELF_COMPONENT_STATUS_OK) {
+		goto error;
+	}
+
+	ret = init_from_params(debug_info_comp, params);
+	if (ret) {
+		BT_LOGE("Cannot configure debug_info component: "
+			"debug_info-comp-addr=%p, params-addr=%p",
+			debug_info_comp, params);
+		goto error;
+	}
+
+	goto end;
+
+error:
+	destroy_debug_info_comp(debug_info_comp);
+	bt_self_component_set_data(
+		bt_self_component_filter_as_self_component(self_comp),
+		NULL);
+
+	if (status == BT_SELF_COMPONENT_STATUS_OK) {
+		status = BT_SELF_COMPONENT_STATUS_ERROR;
+	}
+end:
+	return status;
+}
+
+BT_HIDDEN
+void debug_info_comp_finalize(bt_self_component_filter *self_comp)
+{
+	struct debug_info_component *debug_info =
+		bt_self_component_get_data(
+				bt_self_component_filter_as_self_component(
+					self_comp));
+	BT_LOGD("Finalizing debug_info self_component: comp-addr=%p",
+		self_comp);
+
+	destroy_debug_info_comp(debug_info);
+}
+
+BT_HIDDEN
+bt_self_message_iterator_status debug_info_msg_iter_next(
+		bt_self_message_iterator *self_msg_iter,
+		const bt_message_array_const msgs, uint64_t capacity,
+		uint64_t *count)
+{
+	bt_self_component_port_input_message_iterator *upstream_iterator = NULL;
+	bt_message_iterator_status upstream_iterator_ret_status;
+	struct debug_info_msg_iter *debug_info_msg_iter;
+	struct debug_info_component *debug_info = NULL;
+	bt_self_message_iterator_status status;
+	bt_self_component *self_comp = NULL;
+	bt_message_array_const local_msgs;
+	const bt_message *out_message;
+	uint64_t i, local_count = 0;
+
+	status = BT_SELF_MESSAGE_ITERATOR_STATUS_OK;
+
+	self_comp = bt_self_message_iterator_borrow_component(self_msg_iter);
+	BT_ASSERT(self_comp);
+
+	debug_info = bt_self_component_get_data(self_comp);
+	BT_ASSERT(debug_info);
+
+	debug_info_msg_iter = bt_self_message_iterator_get_data(self_msg_iter);
+	BT_ASSERT(debug_info_msg_iter);
+
+	upstream_iterator = debug_info_msg_iter->msg_iter;
+	BT_ASSERT(upstream_iterator);
+
+	upstream_iterator_ret_status =
+		bt_self_component_port_input_message_iterator_next(
+				upstream_iterator, &local_msgs, &local_count);
+	if (upstream_iterator_ret_status != BT_MESSAGE_ITERATOR_STATUS_OK) {
+		/* No messages were returned. Not necessarily an error. */
+		goto end;
+	}
+
+	/*
+	 * There should never be more received messages than the capacity we
+	 * provided.
+	 */
+	*count = 0;
+	BT_ASSERT(local_count <= capacity);
+
+	for (i = 0; i < local_count; i++) {
+		out_message = handle_message(debug_info_msg_iter, local_msgs[i]);
```

Comment:
`handle_message()` can fail: see all the _can fail_ comments in `handle_*()` functions.

If it fails, don't forget to put the messages that were already created and stored into `local_msgs`.

When this function fails because of no memory, return `BT_SELF_MESSAGE_ITERATOR_STATUS_NOMEM`.